vncviewer/keysym2ucs.c

```diff
@@ -167,8 +167,8 @@ unsigned ucs2keysym(unsigned ucs)
   if (keysym != NoSymbol)
     return keysym;
 
-  /* us the directly encoded 24-bit UCS character */
-  if ((ucs & 0xff000000) == 0)
+  /* ucs is a directly encoded 21-bit Unicode character */
+  if (ucs <= 0x10ffff && ((ucs & 0xfff800) != 0x00d800))
```

Comment:
I guess this is a safety net? Good idea, but I think it is a bit hidden here. Perhaps instead:

```diff
diff --git a/vncviewer/keysym2ucs.c b/vncviewer/keysym2ucs.c
index 6607e3065..ad9b7ebf3 100644
--- a/vncviewer/keysym2ucs.c
+++ b/vncviewer/keysym2ucs.c
@@ -167,6 +167,14 @@ unsigned ucs2keysym(unsigned ucs)
   if (keysym != NoSymbol)
     return keysym;
 
+  /* surrogates? */
+  if (ucs >= 0xd800 && ucs <= 0xdfff)
+    return NoSymbol;
+
+  /* private use? */
+  if (ucs >= 0xe000 && ucs <= 0xf8ff)
+    return NoSymbol;
+
   /* us the directly encoded 24-bit UCS character */
   if ((ucs & 0xff000000) == 0)
     return ucs | 0x01000000;

```

Reply:
Will make more explicit as you suggest, but I will leave the 0x10ffff limit since code points above that values are not valid.